src/Microsoft.AspNet.SignalR.Client/Connection.cs

```diff
@@ -558,17 +558,23 @@ private Task StartTransport()
             return _transport.Start(this, _connectionData, _disconnectCts.Token)
                              .RunSynchronously(() =>
                              {
-                                 // NOTE: We have tests that rely on this state change occuring *BEFORE* the start task is complete
-                                 ChangeState(ConnectionState.Connecting, ConnectionState.Connected);
-
-                                 // Now that we're connected complete the start task that the
-                                 // receive queue is waiting on
-                                 _startTcs.TrySetResult(null);
-
-                                 // Start the monitor to check for server activity
-                                 _lastMessageAt = DateTime.UtcNow;
-                                 _lastActiveAt = DateTime.UtcNow;
-                                 Monitor.Start();
+                                 lock (_stateLock)
+                                 {
+                                     // NOTE: We have tests that rely on this state change occuring *BEFORE* the start task is complete
+                                     if (!ChangeState(ConnectionState.Connecting, ConnectionState.Connected))
+                                     {
+                                         throw new StartException(Resources.Error_ConnectionCancelled, LastError);
+                                     }
+
+                                     // Now that we're connected complete the start task that the
+                                     // receive queue is waiting on
+                                     _startTcs.TrySetResult(null);
```

Comment:
This tcs is dispatching to the thread pool right?

Reply:
Correct. It's a "DispatchingTaskCompletionSource".

Reply:
The change state also calls into user code right?

Reply:
I checked both _startTcs and ChangeState() before submitting this PR to make sure we don't start calling user code under the _stateLock where we weren't before.

The StateChanged event *is* called indirectly by ChangeState(), but it's already being called under the lock: https://github.com/SignalR/SignalR/blob/2c9c833f40cd3c03fdc0de95099bff17118fee43/src/Microsoft.AspNet.SignalR.Client/Connection.cs#L398-L407.